### DISCUSSION TITLE
Suffix cast operator

### DISCUSSION THREAD

**Discussion by TahirAhmadov** [dotnet/csharplang#5447] [Language Ideas]
Is it possible to add an operator to cast which can be appended at the end of the expression being cast?
```cs
string s = (string)o; // current syntax
int l = ((string)f.Foo().Bar).Length; // this quickly becomes difficult to read
int l2 = f.Foo().Bar.CastTo<string>().Length; // what I do in these cases today is use an inlined extension method
int l3 = f.Foo().Bar#string.Length; // using some symbol, be it # or anything else, can get rid of a whole bunch of parenthesis
int l4 = f.Foo().Bar#?string?.Length; // and something like #? can be the equivalent of as
```
Thoughts?

**Comment by huoyaoyuan** [comment]
Generally, the C# team avoids adding new syntax to do the same thing.

**Comment by RikkiGibson** [comment]
The existing cast syntax feels like it's designed to punish you with an extra set of parens for wanting to downcast and immediately access a member. I periodically find myself casting something, then saying: ahh, I need to access a member on this too, I need to go back and wrap it in another set of parens.

Here are some of the places where "casting then accessing a member" occurs within the C# core compiler: https://grep.app/search?q=%5C%28%5C%28%5Cw%2B%5C%29%5Cw%2B%5C%29%5C.&regexp=true&filter[repo.pattern][0]=dotnet/roslyn&filter[path][0]=src/Compilers/CSharp/Portable/ (It would be nice to try and find places where more complex expressions are used as cast operands, but this is still somewhat illustrative.)

```cs
if (someVar.Method() && ((DerivedType)someVar.AnotherMethod()) && someVar.SomethingElse())
{
}
```

I feel like this "papercut" would go away if we made it so a cast could look like a member access a la Go:

```cs
if (someVar.Method() && someVar.(DerivedType).AnotherMethod() && someVar.SomethingElse())
{
}
```

It's less of a "convenience gain" than `?.`, for example, but I think the implementation cost is also lower, and it's relatively straightforward to specify.

**Reply by TahirAhmadov** [reply]
I would be OK with the `.(TypeName)` syntax, too. Perhaps `.(TypeName)?` for `as`?

**Reply by RikkiGibson** [reply]
It seems like `expr.(TypeName)?.Method()` has an intuitive meaning already:
1. cast expr to TypeName
2. if the cast result is not null, call Method on it.

What if it were `expr.(as TypeName)?.Method()`. I think the benefit of something like is only clear when `expr` is something already large and complex, and using the existing form `(expr as TypeName)?.Method()` ends up being a little bit clunky--having to navigate all the way to the beginning of `expr` to add an additional `(`.

All that said, I do not actually know the disposition of any language team members toward the general idea of this. I think that fairly strong evidence would be needed to overcome the "-100 points" hurdle.

**Reply by jnm2** [reply]
Also, `.As<TypeName>()`  might be able to be a half-decent extension method.

**Reply by TahirAhmadov** [reply]
> Also, `.As<TypeName>()` might be able to be a half-decent extension method.

I have both `As` and `CastTo` extension methods. However, they have downsides. First, obviously C# doesn't suggest removal of "useless" casts. Second, whereas `as` and by extension, extension (pun intended) method `As` only work with class types, "casting" and it's `CastTo` counterpart are supposed to work with both class and struct types, but due to the way generics work, I had to make it `where T : class` so it's only useful for class types.

**Reply by TahirAhmadov** [reply]
> It seems like `expr.(TypeName)?.Method()` has an intuitive meaning already:
> 
>     1. cast expr to TypeName
> 
>     2. if the cast result is not null, call Method on it.
> 
> 
> What if it were `expr.(as TypeName)?.Method()`. ...

You are right, of course; I completely forgot about the `?.` operator. I would be OK with your syntax.

**Reply by jnm2** [reply]
> First, obviously C# doesn't suggest removal of "useless" casts.

There is a highly fitting feature that would go along with `.As<T>()`, though, and that is to write an analyzer along with the extension method.

> Second, whereas as and by extensions, extension (pun intended) method As only work with class types, "casting" and it's CastTo counterpart are supposed to work with both class and struct types, but due to the way generics work, I had to make it where T : class so it's only useful for class types.

Here's an option, but it pairs best with an analyzer that errors if non-nullable value types or nullable reference types are used as the generic type argument:

```cs
public static T? As<T>(this object? value)
{
    // Potential `default(T) is null` check
    return value is T ? (T)value : default;
}
```

**Reply by TahirAhmadov** [reply]
@jnm2 thank you for the suggestions. The way I have these extension methods now works more or less. My main motivations for suggesting a language change were 1) make it available to everybody and 2) given how fundamental casting is, it really belongs in the language and not in custom extensions/analyzers/etc. Think about open source projects/etc. - if every team implements their own flavor of casting extension methods, it becomes more difficult to read and collaborate. The same applies even to different members of the same team potentially having disagreements. Language having the feature will serve as a guidance for everybody. Again, this applies to this and not necessarily to many other suggestions because it's so fundamental.

**Reply by jnm2** [reply]
I agree with you. It's always nicer when solutions are in the box. Even so, every team needn't implement their own flavor if a community library+analyzer package rises to prominence.

**Reply by RikkiGibson** [reply]
I find it relevant that this syntax is already "implied" by the conversion operator completion.

![image](https://user-images.githubusercontent.com/5833655/150004877-be47d3fe-885a-4f93-b74d-5f29dc24968e.png)

Accepting the completion actually "pre-wraps" the cast operand, presumably because the tooling expects you to go on and dot something else off of this, and otherwise wants you to run an additional fixer to remove the unnecessary parens.

![image](https://user-images.githubusercontent.com/5833655/150005066-b85c1c61-9039-4049-8eee-95ee58010073.png)

**Reply by RikkiGibson** [reply]
One more reason that it would be better to have this in the language than in a convenience extension is: as far as I am aware, the convenience method can't work with user-defined operators.

Of course, you might introduce some `CastUserDefined` method which uses an `IConvertible` interface with a static conversion operator on it, or something like that. But that doesn't really mesh with any conversions that already exist.

**Reply by TahirAhmadov** [reply]
@RikkiGibson yes and if we were to sum up the downsides of the extension methods: no "value conversions" (like `int` to `long` etc.), no boxing/unboxing of structs (eg. `object` to `int`), and no user-defined conversion operators. And the existing cast/convert syntax (both the `(type)expr` and the `as`) require being wrapped in parenthesis, which is very annoying in many cases.

BTW I think I may have seen a similar proposal for `await`, does that exist or am I mis-remembering?

**Reply by HaloFour** [reply]
@TahirAhmadov 

> BTW I think I may have seen a similar proposal for `await`, does that exist or am I mis-remembering?

Yes, it was rejected: https://github.com/dotnet/csharplang/issues/4076

**Reply by TahirAhmadov** [reply]
@HaloFour actually it wasn't rejected outright; they didn't like the syntax and the narrowness of the solution. Like I said, I don't insist on any particular syntax; if there is a general approach with pipes, I'm all for it. I just hate the parenthesis.
```cs
int l = f.Foo().Bar|>(string).Length; // I would be OK with this, if people want pipes, let them have pipes
```

**Reply by HaloFour** [reply]
@TahirAhmadov 

> actually it wasn't rejected outright

Yeah it was, I created it specifically for an LDM where it was discussed and quickly rejected.  I kind of expected it going into it.

Most of the other proposals for forms for piping/chaining expressions have been sitting and going nowhere since Roslyn went open source.  IMO, it's a problem relatively easily solved by tooling like Intellisense.

**Reply by TahirAhmadov** [reply]
It says: "A more general approach that simplified chaining generally for prefix operators would be more interesting. ... We do like the space of improving await, but we don't think this is the way." It doesn't "outright" to me.
Regarding there being little progress in all these related areas, it's possible those have their own issues? For example, the main use case with piping into first argument of methods has some edge cases which need resolution. I'm not sure those are full fledged proposals that sit completely fleshed out but unimplemented.

**Reply by MgSam** [reply]
@RikkiGibson If this is a serious proposal the LDT might consider maybe you should put it in its own issue / discussion? Very hard to track a popular proposal inside of a discussion where the syntax you propose isn't in the OP.
